torch/nested/_internal/ops.py

```diff
@@ -847,28 +826,75 @@ def is_same_size_default(func, *args, **kwargs):
 
 
 @register_jagged_func(
-    torch.ops.aten.sum.dim_IntList, "self: jt, dim: any?, keepdim: any?, dtype: any?"
+    torch.ops.aten.sum.dim_IntList,
+    "self: jt_all, dim: any?, keepdim: any?, dtype: any?",
 )
 def sum_dim_IntList(func, *args, **kwargs):
-    # sum_dim_IntList can produce a NT or a T depending on whether the ragged dims
-    # are reduced away.
+    """
+    Performs a sum along the provided tensor dimension.
+    Returns a tensor if the ragged dimension is reduced away, else returns a nested tensor.
+    """
     _, new_kwargs = normalize_function(
         func, args=args, kwargs=kwargs, normalize_to_only_use_kwargs=True
     )
     inp = new_kwargs.pop("input")
-    assert inp._ragged_idx == 1
-    new_kwargs["dim"], ragged_reduced_away = _wrap_jagged_dims(
-        inp.dim(), new_kwargs["dim"], "sum"
+    ragged_idx = inp._ragged_idx
+    if ragged_idx != 1:
+        raise RuntimeError("sum(): not supported when ragged_idx != 1 for NestedTensor")
+
+    if inp._lengths is not None:
+        raise RuntimeError(
+            "sum(): not supported for nested tensors with holes, i.e. when lengths is not None, for NestedTensor"
+        )
+
+    new_kwargs["dim"], reduce_on_batch, reduce_on_ragged, reduce_on_non_batch = (
+        _wrap_jagged_dims(
+            inp.dim(),
+            new_kwargs["dim"],
+            "sum",
+            ragged_idx,
+        )
     )
 
-    if not ragged_reduced_away:
+    # invalid reductions
+    batch = reduce_on_batch and not reduce_on_ragged and not reduce_on_non_batch
+    batch_and_non_batch = (
+        reduce_on_batch and not reduce_on_ragged and reduce_on_non_batch
+    )
+    ragged_and_non_batch = (
+        not reduce_on_batch and reduce_on_ragged and reduce_on_non_batch
+    )
+
+    if batch or batch_and_non_batch:
+        raise RuntimeError(
+            "sum(): not supported along a batch dimension for NestedTensor"
+        )
+
+    if ragged_and_non_batch:
+        raise RuntimeError(
+            "sum(): not supported along a ragged and non-batch dimension for NestedTensor"
+        )
```

Comment:
nit: I find this a bit tricky to read, but it seems like these are the cases we care about:
1. raggedness is reduced away -> return dense
    * not reducing over batch dim -> need to read offsets to determine batch items
        * error for non-batch as unsupported
        * call `_jagged_to_padded_dense_forward()` otherwise
    * reducing over batch dim -> no need to read offsets
        * call `func` directly on `values` (NB: avoid duplicate dims in the dim_list arg)
2. raggedness is not reduced away -> return NJT
    * error if trying to reduce over the batch dim
    * call `func` directly on `values`

maybe structuring like this (with comments) could make it easier to follow, wdyt?

Reply:
i agree, i think this is a good point on readability! refactored my code to make it easier to understand, with comments clearly explaining what is currently supported and what isn't. thank you!